### DISCUSSION TITLE
Define css variable inside dynamic styles

### DISCUSSION THREAD

**Discussion by tienlx97** [facebook/stylex#134] [Q&A] [answered]
- I declare `'--color'` variable inside `stylex.create` with dynamic styles . It seems work!
- Instead of declaring `--color`, it declares `----color`.
- This is my code: 

```jsx

import * as stylex from '@stylexjs/stylex';

const styles = stylex.create({
  color: (color?: string) => ({
    color: 'var(--color,revert)',
    '--color': color ?? 'initial',
  }),

  propsFromParent_01: {
    display: 'block',
    transitionDuration: 'var(--fds-fast)',
    transitionProperty: 'color,fill,stroke',
    transitionTimingFunction: 'var(--fds-soft)',
  },

  propsFromParent_02: {
    display: 'block',
  },
});

export default function Home() {
  const color = 'var(--primary-icon)';

  return (
    <div
      {...stylex.props([
        color && styles.color(color),
        // pass this from parent through props
        [styles.propsFromParent_01, !color && styles.propsFromParent_02],
      ])}
    >
      Hello
    </div>
  );
}

```

- This is compiled code : 
![image](https://github.com/facebook/stylex/assets/25848009/3c646b06-d30d-4d99-8a5a-4a8457d5e036)

- This is sandbox: https://stackblitz.com/edit/stylex-next-pb8mt5?file=app%2Fpage.tsx

**Comment by nmn** [comment, marked as answer]
You're doing the thing that StyleX does for you automatically. Please use this code instead:

```ts
import * as stylex from '@stylexjs/stylex';

const styles = stylex.create({
  color: (color?: string) => ({ color }),

  propsFromParent_01: {
    display: 'block',
    transitionDuration: 'var(--fds-fast)',
    transitionProperty: 'color,fill,stroke',
    transitionTimingFunction: 'var(--fds-soft)',
  },

  propsFromParent_02: {
    display: 'block',
  },
});

export default function Home({color, style}) {

  return (
    <div
      {...stylex.props([
        color && styles.color(color),
        style,
      ])}
    >
      Hello
    </div>
  );
}
```

Note, you should never have to write variables as strings directly in StyleX. It's core goal to not have to deal with "magic strings" like that. Everything should be a reference with StyleX.

**Comment by HQ92** [comment]
I had a question here with regards to support for stylex being able to write out custom css variables, ie magic strings for certain use cases which can be used with globalvars to use stylex to manage themes for other plugins and components systems such as shadcn/ui. It needs you to define certain css variables with values to customise form elements, what you'd really want to do is define those colours and values inside stylex, and then to simply map the shadcn variables to stylex generated variables to have everything controlled from one location?

The current method would require duplicating code between the imported css file where you could set css variables for shadcn in the current normal way with css, and then you'd treat stylex as a completely separate system for everything else, which isn't that preferable.

**Reply by nmn** [reply]
I'm not sure I understood your question, but it seems like you want to be able to set values for named variables that are defined outside of StyleX.

This can be done with `stylex.create`. You can set styles for arbitrary `--var-name` properties in addition to CSS properties like `color` and `margin`. This is an escape hatch though and you will get ESLint errors. You can suppress the lint errors if this helps unblock you.

**Reply by HQ92** [reply]
I tried it and it works, thank you. It does still show an error when doing so with eslint however, is this something you plan to support more concretely?

**Reply by nmn** [reply]
This is already fully supported, but the ESLint rule is made for the common case.
